### PR TITLE
Java: Update ToolCallBehavior

### PR DESCRIPTION
**Pull request by milderhc**
### Contribution Checklist

- [ ] The code builds clean without any errors or warnings
- [ ] The PR follows the [SK Contribution Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md) and the [pre-submission formatting script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts) raises no violations
- [ ] All unit tests pass, and I have added new tests where possible
- [ ] I didn't break anyone :smile:

### CODE REVIEW

File: java/semantickernel-api/src/main/java/com/microsoft/semantickernel/orchestration/ToolCallBehavior.java [hunk 1]
```diff
@@ -59,14 +59,15 @@ public UnmodifiableToolCallBehavior unmodifiableClone() {
...
+     * to the model unless specific functions have been enabled via {@code enableFunctions}.
+     * <p>
+     * By default, all kernel functions are disabled.
      *
-     * @param enable Whether to enable kernel functions.
      * @return This ToolCallBehavior.
      */
-    public ToolCallBehavior enableKernelFunctions(boolean enable) {
-        this.kernelFunctionsEnabled = enable;
+    public ToolCallBehavior enableKernelFunctions() {
```
Comment (johnoliver) [inline java/semantickernel-api/src/main/java/com/microsoft/semantickernel/orchestration/ToolCallBehavior.java:69]: Can we update ToolCallBehaviour to construct via builder pattern and make fields final

File: java/semantickernel-api/src/main/java/com/microsoft/semantickernel/orchestration/ToolCallBehavior.java [hunk 2]
```diff
@@ -201,33 +101,135 @@ public int getMaximumAutoInvokeAttempts() {
...
+         */
+        public Builder autoInvoke(boolean enable) {
+            return withMaximumAutoInvokeAttempts(enable ? DEFAULT_MAXIMUM_AUTO_INVOKE_ATTEMPTS : 0);
         }
 
-        @Override
-        public final ToolCallBehavior enableKernelFunctions(boolean enable) {
-            throw new UnsupportedOperationException("unmodifiable instance of ToolCallBehavior");
+        /**
+         * Require or not require a function to be called.
```
[the remark this answers is not in the document]
Comment (milderhc) [reply java/semantickernel-api/src/main/java/com/microsoft/semantickernel/orchestration/ToolCallBehavior.java:145]: Only one required function is possible.

[the remark this answers is not in the document]
Comment (dsgrieve) [reply java/semantickernel-api/src/main/java/com/microsoft/semantickernel/orchestration/ToolCallBehavior.java:145]: Need to state that in the javadoc.

File: java/semantickernel-api/src/main/java/com/microsoft/semantickernel/orchestration/ToolCallBehavior.java [hunk 3]
```diff
@@ -201,33 +101,135 @@ public int getMaximumAutoInvokeAttempts() {
...
+        public Builder requireFunction(KernelFunction<?> function) {
+            requiredFunction = function;
+            return withMaximumAutoInvokeAttempts(1);
         }
 
-        @Override
-        public final ToolCallBehavior autoInvoke(boolean enable) {
-            throw new UnsupportedOperationException("unmodifiable instance of ToolCallBehavior");
+        /**
+         * Enable a function.
```
[the remark this answers is not in the document]
Comment (milderhc) [reply java/semantickernel-api/src/main/java/com/microsoft/semantickernel/orchestration/ToolCallBehavior.java]: The last statement is not like that. If you call `enableFunctions` those functions will be passed to the model and may be called. `enableKernelFunctions` simply enables all of the kernel functions to be passed. `enableFunctions` only enables a subset.

File: java/semantickernel-api/src/main/java/com/microsoft/semantickernel/orchestration/ToolCallBehavior.java [hunk 4]
```diff
@@ -201,33 +101,135 @@ public int getMaximumAutoInvokeAttempts() {
...
+         *
+         * @param maximumAutoInvokeAttempts The maximum number of attempts.
+         * @return This ToolCallBehavior.
+         */
+        public Builder withMaximumAutoInvokeAttempts(int maximumAutoInvokeAttempts) {
+            if (maximumAutoInvokeAttempts < 0) {
+                throw new SKException(
+                        "The maximum auto-invoke attempts should be greater than or equal to zero.");
+            }
+            if (requiredFunction == null) {
```
[the remark this answers is not in the document]
Comment (milderhc) [reply java/semantickernel-api/src/main/java/com/microsoft/semantickernel/orchestration/ToolCallBehavior.java]: If the required function is set, there is no need to send all of the other enabled functions from the kernel as the model will try to call only that function.

File: java/semantickernel-api/src/main/java/com/microsoft/semantickernel/orchestration/ToolCallBehavior.java [hunk 5]
```diff
@@ -201,33 +101,135 @@ public int getMaximumAutoInvokeAttempts() {
...
+         * to the model. If the maximum number of attempts is reached, the model will be notified that
+         * the function could not be invoked. The default maximum number of attempts is 5.
+         *
+         * @param maximumAutoInvokeAttempts The maximum number of attempts.
+         * @return This ToolCallBehavior.
+         */
+        public Builder withMaximumAutoInvokeAttempts(int maximumAutoInvokeAttempts) {
+            if (maximumAutoInvokeAttempts < 0) {
+                throw new SKException(
+                        "The maximum auto-invoke attempts should be greater than or equal to zero.");
```
[the remark this answers is not in the document]
Comment (milderhc) [reply java/semantickernel-api/src/main/java/com/microsoft/semantickernel/orchestration/ToolCallBehavior.java]: Removed.

File: java/semantickernel-api/src/main/java/com/microsoft/semantickernel/orchestration/ToolCallBehavior.java [hunk 6]
```diff
@@ -201,33 +101,135 @@ public int getMaximumAutoInvokeAttempts() {
...
+         * Require or not require a function to be called.
+         * If a function is required, it will the only function to be passed to the model and be called.
+         * By default, no function is required.
+         *
+         * @param function The function to require or not require.
+         * @return This ToolCallBehavior.
+         */
+        public Builder requireFunction(KernelFunction<?> function) {
+            requiredFunction = function;
+            return withMaximumAutoInvokeAttempts(1);
```
[the remark this answers is not in the document]
Comment (milderhc) [reply java/semantickernel-api/src/main/java/com/microsoft/semantickernel/orchestration/ToolCallBehavior.java]: There was no need to have the ability to set the maximum auto invoke attempts having the default maximum at a low value. I removed this.

File: java/semantickernel-api/src/main/java/com/microsoft/semantickernel/orchestration/ToolCallBehavior.java [hunk 7]
```diff
@@ -21,124 +24,21 @@ static String getKey(@Nullable String pluginName, String functionName) {
         return String.format("%s%s%s", pluginName, FUNCTION_NAME_SEPARATOR, functionName);
     }
 
-    private int maximumAutoInvokeAttempts;
-    private boolean kernelFunctionsEnabled;
-
+    private final int maximumAutoInvokeAttempts;
```
[the remark this answers is not in the document]
Comment (milderhc) [reply java/semantickernel-api/src/main/java/com/microsoft/semantickernel/orchestration/ToolCallBehavior.java:26]: I think you have a different concept of enable kernel functions. `enableKernelFunctions` is a shortcut to enable all of them to be passed to the model. At the end, all the functions that may be passed to the model belong to the kernel, so they are already _enabled_ (i.e. available) to be passed, the user decides which set of those should actually be sent to the model.

**Review by johnoliver — APPROVED** [review]

**Pull request merged**